Clean up user data on logout

miankhalid opened the pull request:
### Description

[MA-2515](https://openedx.atlassian.net/browse/MA-2515)

The PR does 4 main things:

1. Nuke user data except videos and downloads database.
2. Runs migrations of the database to hash the usernames to their SHA1 counterparts.
3. Runs migrations to change the legacy directory structure of how we store videos and their subtitles.
4. Silently fetch and cache course structure of the courses we have downloaded videos of.

### Notes
- To properly test this PR, first make a build out of current master and then overwrite the build from this PR to view the migrations in action.
- Results of migrations are more visible if testing is being done on the emulator (cuz we can see whatever files we want to see from Android Device Monitor).

### Testing
- [x] Manual tests this PR should pass

Following tests should pass on this PR:

1. Make a build from master, download some videos and then make a build from this branch. The directory that stores videos should be updated i.e. previously the videos were stored like this `sdcard/org.edx.mobile/staff/(SHA1_hash_of_video_name)`, now it should appear as `sdcard/org.edx.mobile/videos/(SHA1_of_username)/(SHA1_hash_of_video_name)`.
2. `srtFiles` has been renamed to `subtitles` and has been moved inside `videos` folder.
3. The username field in the database should now have SHA1 representation of the username.
4. After logout, in the app's directory only `databases/download.db` should be there and in the app's external directory only the `videos` folder should remain.

- [x] Database migrations are backwards-compatible

Database has been upgraded from version 5 to version 6.

### Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [x] Code review: @1zaman 
- [x] Code review: @BenjiLee 
- [x] Code review: @mdinino  
- [ ] Code review: @bguertin

bguertin commented:
Hmmm, If we are migrating the location of all the video files, it might be a good time to move them to the correct folder (our app's data folder) instead of an arbitrary place on the SD card, so that they are properly deleted when uninstalling the edX app.

Not necessary for this PR of course, just thinking we might want to do one migration and not two :confused:

miankhalid commented:
> Hmmm, If we are migrating the location of all the video files, it might be a good time to move them to the correct folder (our app's data folder) instead of an arbitrary place on the SD card, so that they are properly deleted when uninstalling the edX app.

Well we could certainly do that, but uninstalling or clearing the data, cleans up the app's whole directory on SD Card so every single video and subtitles are properly deleted.

bguertin commented:
> uninstalling or clearing the data, cleans up the app's whole directory on SD Card so every single video and subtitles are properly deleted.

Hm, right, my bad. The actual issue, now I'm remember, is that we require the storage permission because we are technically modifying files outside our apps `files` directory. If we migrated all our users to use the "files" directory, we can remove the requirement for having the "files" permission.

miankhalid commented:
> If we migrated all our users to use the "files" directory, we can remove the requirement for having the storage permission.

Wouldn't that migration also require the write and read permission. Plus, in our future plans we might allow user to store a video on a custom location, bam! write permission again 😉

mdinino commented on VideoLocker/src/main/java/org/edx/mobile/http/cache/CacheManager.java:
```diff
@@ -34,16 +32,13 @@ public CacheManager(Context context) {
 
-    public boolean has(String url) throws NoSuchAlgorithmException,
-            UnsupportedEncodingException {
+    public boolean has(String url) {
```
It looks like url needs to be non-null, correct? (Or else an NPE will occur). If so, can you mark accordingly? In fact, even if it is nullable, can you mark accordingly?

mdinino commented on VideoLocker/src/main/java/org/edx/mobile/http/cache/CacheManager.java:
```diff
@@ -34,16 +32,13 @@ public CacheManager(Context context) {
-    public void put(String url, String response)
-            throws NoSuchAlgorithmException, UnsupportedEncodingException,
-            IOException {
+    public void put(String url, String response) throws IOException {
```
same comment here, can you mark as @Nonnull or @Nullable as appropriate?

mdinino commented on VideoLocker/src/main/java/org/edx/mobile/http/cache/CacheManager.java:
```diff
@@ -52,7 +47,7 @@ public void put(String url, String response)
     }
 
-    public String get(String url) throws IOException, NoSuchAlgorithmException {
+    public String get(String url) throws IOException {
```
@Nonnull?

mdinino commented on VideoLocker/src/main/java/org/edx/mobile/module/db/impl/IDatabaseImpl.java:
```diff
@@ -32,7 +33,9 @@ public IDatabaseImpl(Context context, LoginPrefs loginPrefs) {
     private String username() {
-        return loginPrefs.getUsername();
+        final String username = loginPrefs.getUsername();
+        if (username != null) return Sha1Util.SHA1(username);
```
nit pick: Can we make this one return statement?  Something like:
return (username != null) ? Sha1Util.SHA1(username) : null;

miankhalid replied:
sure. Will do.

1zaman replied:
My preference here would actually be to just change the `username` variable itself:

```` java
if (username != null) {
    username = Sha1Util.SHA1(username);
}
return username;
````

This seems to be the clearest way to write it. But I'm fine with however you want to do it.

mdinino commented on VideoLocker/src/main/java/org/edx/mobile/player/TranscriptManager.java:
```diff
@@ -31,29 +30,20 @@
      */
-    public boolean has(String url) throws NoSuchAlgorithmException,
-    UnsupportedEncodingException {
+    public boolean has(String url) {
```
@Nonnull?

mdinino commented on VideoLocker/src/main/java/org/edx/mobile/player/TranscriptManager.java:
```diff
@@ -84,26 +70,20 @@ public void put(String url, String response)
-            return cache;
-        }catch(Exception e){
-            logger.error(e);
+    public String get(String url) throws IOException {
```
@Nonnull?

mdinino commented on VideoLocker/src/main/java/org/edx/mobile/player/TranscriptManager.java:
```diff
@@ -112,34 +92,24 @@ public String get(String url) throws IOException, NoSuchAlgorithmException {
-            return in;
-        }catch(Exception e){
-            logger.error(e);
+    public InputStream getInputStream(String url) throws IOException {
```
@Nonnull?

mdinino commented on VideoLocker/src/main/java/org/edx/mobile/player/TranscriptManager.java:
```diff
@@ -182,58 +148,22 @@ public void downloadTranscriptsForVideo(TranscriptModel transcript){
-            } catch (UnsupportedEncodingException e) {
-                logger.error(e);
-            }
+            startTranscriptDownload(transcript.chineseUrl);
```
There seems to be a lot of copy/paste code here. If it's not too much trouble can we move all if this checking/downloading for each language into one helper function? I don't mind if we manually call invoke each language, but it seems redundant to check each language for non null and then callStartTranscriptDownload on each one

mdinino commented on VideoLocker/src/main/java/org/edx/mobile/util/AppConstants.java:
```diff
@@ -1,12 +1,17 @@
 
 
-public class AppConstants {
+public final class AppConstants {
```
My suggestion would be to make this class final and static (unless ther eis a good reason not to)

bguertin replied:
This class is already final and static.

miankhalid replied:
Lemme make that an enum class, which will won't allow it to be extended nor an object of it to be instantiated.

mdinino commented on VideoLocker/src/main/java/org/edx/mobile/util/AppConstants.java:
```diff
@@ -1,12 +1,17 @@
     public static final double MILLISECONDS_PER_SECOND = 1000.00;
+
+    public static final class Folders {
+        public static final String VIDEOS = "videos";
```
I believe static final variables are redundant here.

bguertin replied:
@mdinino Could you elaborate? I'm not sure what you mean by that.

mdinino replied:
Actually, scrratch that. I was wrong. Deleting comment.

mdinino commented:
Done with pass. I had minor suggestions, but other than those it looks great.

bguertin commented:
> Plus, in our future plans we might allow user to store a video on a custom location, bam! write permission again

But, we could at least save them in the default location without asking for permission. Anyway, forget I said anything, we can just move them again if we have to :package: 

I agree about the null/nonnull annotations, but I wouldn't worry about annotating all the existing methods that weren't really affected by this PR, unless it is straightforward (no need to increase scope for that).

BenjiLee commented on VideoLocker/src/test/java/org/edx/mobile/test/UtilTests.java:
```diff
@@ -32,4 +37,10 @@ public void testHostAndUrls() throws Exception {
     }
+
+    @Test
+    public void testSha1Encryption() throws UnsupportedEncodingException, NoSuchAlgorithmException {
```
Tests :3

miankhalid replied:
oops, this test shouldn't now be throwing those exceptions. I'll update it when the review passes complete.

1zaman replied:
I think it would be better to have separate test suites for each utility class for better organization (both for the reader and the test code itself).

miankhalid commented on VideoLocker/src/main/java/org/edx/mobile/util/FileUtil.java:
```diff
@@ -38,4 +40,25 @@
+     * @param fileOrDirectory The file/folder that needs to be deleted.
+     * @param exceptions      Name of the file/folder that needs to be skipped while deletion.
+     */
+    public static void deleteRecursive(File fileOrDirectory, List<String> exceptions) {
```
oops, lemme fix that.

1zaman replied:
Maybe the method name could be changed to `deleteRecursively()` instead.

miankhalid commented:
@mdinino most of the refactoring reviews that you gave are related to this commit 763026f, which isn't actually what the PR is addressing at core, so, counting them as out of scope.

BenjiLee commented on VideoLocker/src/main/java/org/edx/mobile/module/prefs/PrefManager.java:
```diff
@@ -404,4 +412,11 @@ public void setLastCourseStructureFetch(String courseId, long timestamp) {
         public static final String BACKEND_GOOGLE = "google-oauth2";
     }
+
+    public static void nukeSharedPreferences() {
```
Docstring would be helpful to clarify what we mean by nuke. 

Looks like nuke in our case is just emptying out the user data instead of deleting the shared_prefs folder.

1zaman replied:
Why not just iterate through the 'shared_prefs' directory and delete all the pref files, instead of just emptying them?

BenjiLee commented:
I think you need to rebase before tests can run on travis.

👍

mdinino commented:
👍

miankhalid commented:
All review comments have been addressed and branch has been rebased on master. Just need you review @1zaman before I merge this PR.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/util/FileUtil.java:
```diff
@@ -38,4 +41,25 @@
+
+    /**
+     * Deletes a single file and/or recursively deletes files/folders within a folder before
+     * deleting it.
```
The 'and/or' should technically be just 'or'. Or the description could be simplified to e.g. just "Deletes a file or directory and all contents recursively.".

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/util/FileUtil.java:
```diff
@@ -38,4 +41,25 @@
+     * Deletes a single file and/or recursively deletes files/folders within a folder before
+     * deleting it.
+     *
+     * @param fileOrDirectory The file/folder that needs to be deleted.
```
The '/' should be replaced with 'or'. Applies to the documentation of the next param as well.

Also, 'directory' is a technically more accurate terminology than 'folder'. This applies to the whole javadoc here.

1zaman replied:
I think the param name can be just `file` here instead of `fileOrDirectory`, as the `File` object is known to be generic in Java.

miankhalid replied:
Its more verbose the way it is.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/util/FileUtil.java:
```diff
@@ -38,4 +41,25 @@
+     * deleting it.
+     *
+     * @param fileOrDirectory The file/folder that needs to be deleted.
+     * @param exceptions      Name of the file/folder that needs to be skipped while deletion.
```
The '/' here should be replaced with 'or'.

Also note that 'directory' is a technically more accurate terminology than 'folder'.

1zaman replied:
Also, "Name" and "file/folder" should both at least be plural.

1zaman replied:
To be precise, the `exceptions` parameter should be either a list of fully qualified filenames (whether absolute or relative to the file parameter), or (even better) a list of actual `File` objects.

1zaman replied:
You shouldn't be sloppy when writing utility code, and this change isn't too hard to implement. But the final decision is up to you.

miankhalid replied:
Keeping it as is.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/util/FileUtil.java:
```diff
@@ -38,4 +41,25 @@
+                }
+            }
+        }
+        fileOrDirectory.delete();
```
There is a warning here for ignoring the result of this method. I suppose that behaviour is fine in this case, since we don't want to short-circuit the whole recursive operations on encountering one error. You can suppress it by adding `//noinspection ResultOfMethodCallIgnored` before this line.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/util/SecurityUtil.java:
```diff
@@ -0,0 +1,50 @@
+                File[] children = appDir.listFiles();
+                for (int size = children.length, i = 0; i < size; i++) {
+                    FileUtil.deleteRecursive(children[i],
+                            Collections.singletonList(DbStructure.NAME));
```
Why not do the `deleteRecursive()` call on the root directory itself, instead of iterating over all it's children? As long as there is an undeleted file in it, the root shouldn't actually be deleted itself.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/util/SecurityUtil.java:
```diff
@@ -0,0 +1,50 @@
+    private static final Logger logger = new Logger(SecurityUtil.class);
+
+    public static void clearUserData(@NonNull Context context) {
+        // Clear the App's directory
```
This should say "Clear the data directory" to be unambiguous.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/util/SecurityUtil.java:
```diff
@@ -0,0 +1,50 @@
+        PackageManager packageManager = context.getPackageManager();
+        try {
+            PackageInfo packageInfo = packageManager.getPackageInfo(context.getPackageName(), 0);
+            File appDir = new File(packageInfo.applicationInfo.dataDir);
```
Again, this should probably be named `dataDir` instead.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/util/SecurityUtil.java:
```diff
@@ -0,0 +1,50 @@
+        }
+
+        // Now clear the App's external storage directory
+        File externalDir = new File(context.getExternalFilesDir(null).getParent());
```
I suggest renaming `externalDir` to `externalFilesDir` as well.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/util/SecurityUtil.java:
```diff
@@ -0,0 +1,50 @@
+            File appDir = new File(packageInfo.applicationInfo.dataDir);
+            if (appDir.exists() && appDir.isDirectory()) {
+                File[] children = appDir.listFiles();
+                for (int size = children.length, i = 0; i < size; i++) {
```
This can be simplified by converting it to a foreach loop. Note that using foreach loop in arrays don't even have the (negligible) performance penalty of having to create an `Iterator`.

```` java
for (final File child : appDir.listFiles())
````

miankhalid replied:
Sold.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/view/NavigationFragment.java:
```diff
@@ -231,6 +233,7 @@ public void onClick(View v) {
             @Override
             public void onClick(View v) {
                 environment.getRouter().forceLogout(getActivity(), environment.getSegment(), environment.getNotificationDelegate());
+                SecurityUtil.clearUserData(MainApplication.application);
```
This should be moved to the `forceLogout()` method implementation itself, rather than hardcoded inside a specific click listener.

miankhalid replied:
There was a requirement to only clear a user's data when he explicitly logs out i.e. using the Logout button. Another case of automatic logout is the token expiry case.

Javadoc of forceLogout function:
`````java
    /**
     * this method can be called either through UI [ user clicks LOGOUT button],
     * or programmatically
     */

1zaman replied:
Hmm, good point. It seems that my implementation for invalidating the refresh token from there in #800 was incorrect as well. Maybe there should be some other decoupled method for implementing manual logout that's invoked by the actual click listener on the logout button, which would then have the specific logic for manual logout, and then call through to this `forceLogout()` method.

miankhalid replied:
Umm, invalidating tokens didn't have the special requirement for it to work only when a user interaction happens, so, I guess its correct as is.

1zaman replied:
The token invalidation was also supposed to be only performed upon manual logout. If an automatic logout is performed due to an authorization issue, then the invalidation would seem to be meaningless (though I suppose there's no harm in doing so).

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/util/Sha1Util.java:
```diff
@@ -1,24 +1,28 @@
-     * @throws NoSuchAlgorithmException
-     * @throws UnsupportedEncodingException
+     * @param text The plain text to encrypt.
+     * @return SHA1 hash of the given text or the plain text if encryption failed.
```
Encryption is not synonymous with hashing.

miankhalid replied:
good catch.

1zaman replied:
Note that the 'encryption' phrase is present in a lot of places, including variables and method names. These should all be updated, I suppose.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/util/SecurityUtil.java:
```diff
@@ -0,0 +1,50 @@
+package org.edx.mobile.util;
+
+
```
Extra newline.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/module/db/DbStructure.java:
```diff
@@ -8,9 +8,12 @@
+     * Updated to Version 4 to add flag to indicate that video is only available for web
+     * Updated to Version 5 to create a new table to record learning history for assessment
+     * Updated to Version 6 to swap every occurrence of username field to its SHA1 counterpart
+     */
```
I don't think this should be the official javadoc of the `VERSION` constant (the `/**` at the start makes it a javadoc); if it's to be in a javadoc then it should probably be in the documentation of the `DbHelper#onUpgrade()` method.

1zaman replied:
The latter option seems to be the more proper.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/module/db/impl/DbHelper.java:
```diff
@@ -14,17 +20,18 @@
 class DbHelper extends SQLiteOpenHelper {
-
     private SQLiteDatabase sqliteDb;
+    private Context mContext;
```
Do we need to use these 'm' and 's' field prefixes? We've not been using them previously, and I don't see any need for them, since IDEs highlight their scopes anyway.

miankhalid replied:
ok, I'll remove it.

miankhalid commented:
@1zaman do notify by an ending comment when ur pass finishes, so that I can make the said changes and make this PR ready for another/final pass.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/module/db/impl/DbHelper.java:
```diff
@@ -109,12 +116,60 @@ public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
 
+            if (oldVersion < 6) {
+                Cursor cursor = db.query(false, DbStructure.Table.DOWNLOADS,
+                        new String[]{DbStructure.Column.ID, DbStructure.Column.USERNAME,
```
There should be a space between `[]{`, and preferably after it as well.

miankhalid replied:
Autoformat doesn't do it, so I guess, we shouldn't too.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/module/db/impl/DbHelper.java:
```diff
@@ -109,12 +116,60 @@ public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
+                        new String[]{DbStructure.Column.ID, DbStructure.Column.USERNAME,
+                                DbStructure.Column.FILEPATH}, null, null, null, null, null, null);
+                if (cursor != null) {
+                    final File appExternalDir = mContext.getExternalFilesDir(null).getParentFile();
```
This statement should either be moved to before the query, or inside the try block; we should guarantee the closing of the `Cursor` in all cases after a successful query.

Also note that the `getExternalFilesDir()` method can return `null` if the external storage is not mounted. So I suppose you should check this before the query, and only perform the migration if it returns a non-`null` value.

miankhalid replied:
The placement is currently perfect, the only issue is that appExternalDir shouldn't be `null` for a smooth traversal of the following code. Fret not, I am fixing it.
More info here: https://github.com/edx/edx-app-android/pull/797/files#r76778034

1zaman replied:
No it's not perfect; as I said, it needs to be either in the `try` block, or before the query.

miankhalid replied:
What if refactored this statement to never return a `null` i.e. we ensured that we won't get NPE on this statement ?

1zaman replied:
The method may throw an unexpected exception or error. Since you're using a `try`/`finally` pattern for closing the `Cursor` (on code that doesn't throw any checked exceptions either), any code after getting the `Cursor` should be in the `try` block. Otherwise you might as well get rid of the `try`/`finally` construct.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/module/db/impl/DbHelper.java:
```diff
@@ -109,12 +116,60 @@ public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
+                                    appExternalDir.getAbsolutePath() + "/" + username,
+                                    appExternalDir.getAbsolutePath() + "/"
+                                            + AppConstants.Folders.VIDEOS + "/"
+                                            + encryptedUsername);
```
Since both of these paths should constitute the whole string, how about just creating a new string instead of replacing?

```` java
if (!(appExternalDir.getAbsolutePath() + '/' + username)
        .equals(filePath)) {
    continue;
}
final String newFilePath = appExternalDir.getAbsolutePath() + "/"
        + AppConstants.Folders.VIDEOS + "/"
        + encryptedUsername;
````

1zaman replied:
Also, do we know that the `getExternalFilesDir()` method would return the same media path that was constructed manually previously via the global `Environment.getExternalStorageDirectory()` method? The documentation on the former states that the returned path may change over time if different shared storage media is inserted, but the latter's documentation only states that it's path may change between different platform versions. If there may be a discrepancy, then we should check for the previous path and replace it with the latter (possibly via regex).

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/module/db/impl/DbHelper.java:
```diff
@@ -109,12 +116,60 @@ public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
+                            final File previousFolder = new File(appExternalDir, username);
+                            if (previousFolder.exists()) {
+                                final File newFolder = new File(appExternalDir,
+                                        AppConstants.Folders.VIDEOS + "/" + encryptedUsername);
```
Maybe the 'folder' in the variable names here could be replaced with 'directory' instead?

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/module/db/impl/DbHelper.java:
```diff
@@ -109,12 +116,60 @@ public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
+                            if (previousFolder.exists()) {
+                                final File newFolder = new File(appExternalDir,
+                                        AppConstants.Folders.VIDEOS + "/" + encryptedUsername);
+                                if (!newFolder.exists()) newFolder.mkdirs();
```
The `mkdirs()`method doesn't throw an exception, so you don't need to add the `exists()` check on it. That would make this whole logic an atomic filesystem operation as well.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/module/db/impl/DbHelper.java:
```diff
@@ -109,12 +116,60 @@ public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
+                                final File newFolder = new File(appExternalDir,
+                                        AppConstants.Folders.VIDEOS + "/" + encryptedUsername);
+                                if (!newFolder.exists()) newFolder.mkdirs();
+                                previousFolder.renameTo(newFolder);
```
Does the renaming still work after the destination directory has already been created? In any case, I suppose you should check it's return value, and skip the entry if it reports failure.

1zaman replied:
I suppose the rename works if the new path is an empty directory. Anyway, as I mentioned, you should check the return value of both of these operations, and skip the iteration if either of them fails.

```` java
if (!((newFolder.mkdirs() || newFolder.exists()) && previousFolder.renameTo(newFolder))) {
    continue;
}
````

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/module/db/impl/DbHelper.java:
```diff
@@ -109,12 +116,60 @@ public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
+                                            + AppConstants.Folders.VIDEOS + "/"
+                                            + encryptedUsername);
+
+                            // First update the folder names
```
Only one folder is being updated, so 'names' should be singular I suppose. And since the path is also being updated, you should probably add this info as well.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/module/db/impl/DbHelper.java:
```diff
@@ -109,12 +116,60 @@ public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
+                                previousFolder.renameTo(newFolder);
+                            }
+
+                            // Then update the database rows
```
Again, only one database row is being updated here, so this should be singular.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/module/db/impl/DbHelper.java:
```diff
@@ -109,12 +116,60 @@ public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
+                            updatedValues.put(DbStructure.Column.USERNAME, encryptedUsername);
+                            updatedValues.put(DbStructure.Column.FILEPATH, newFilePath);
+                            db.update(DbStructure.Table.DOWNLOADS, updatedValues,
+                                    DbStructure.Column.ID + "=" + id, null);
```
You should bind the `id` parameter by passing it as a string array (with a single item) in the last parameter here, so that the statement doesn't change each time, and can thus be compiled to a prepared statement by the SQLite session implementation. Note that you can get it directly as a string from the `Cursor` itself if you want.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/module/db/impl/DbHelper.java:
```diff
@@ -109,12 +116,60 @@ public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
+                    } finally {
+                        cursor.close();
+                    }
+                    // Now migrate the subtitles folder
```
There should be an empty line separator before this new task I guess.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/module/db/impl/DbHelper.java:
```diff
@@ -109,12 +116,60 @@ public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
+                        final File newSrtFolder = new File(appExternalDir,
+                                AppConstants.Folders.VIDEOS + "/" + AppConstants.Folders.SUBTITLES);
+                        if (!newSrtFolder.exists()) newSrtFolder.mkdirs();
+                        previousSrtFolder.renameTo(newSrtFolder);
```
The same two comments on renaming the users directories applies here as well. Also, shouldn't the subtitles go into the users directory as well (along with the actual videos)?

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/module/db/impl/DbHelper.java:
```diff
@@ -109,12 +116,60 @@ public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
             logger.debug("Database upgraded from " + oldVersion + " to " + newVersion);
         }catch(Exception e){
             logger.error(e);
         }
```
Should we be catching and discarding all exceptions when attempting to update the database schema? That may result in an inconsistent state, where the version number is updated but the schema still reflects that of an older version, which may cause issues in the future. All the database calls in the codebase seem to be written to catch exceptions anyway, which would seem to be the proper point of failure.

bguertin replied:
We definitely should not be discarding the exception. If we do, the app will consider the upgrade successful even if it fails, and will never try to upgrade it again for this database version.

bguertin replied:
We aren't even logging this error to crashlytics; we could have countless failed upgrades in the wild right now :(

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/module/prefs/UserPrefs.java:
```diff
@@ -51,20 +52,21 @@ public boolean isDownloadOverWifiOnly() {
-        File packDir = new File(downloadsDir, context.getPackageName());
-        File edxDir = new File(packDir, profile.username);
-        edxDir.mkdirs();
+        File videosDir = new File(context.getExternalFilesDir(null).getParent(), AppConstants.Folders.VIDEOS);
```
Note that the documentation for `getExternalFilesDir()` states that it returns `null` if shared storage is not available, while the `getExternalStorageDirectory()` documentation states that it will return the `File` representing the directory path regardless of whether it's currently accessible.

Possible fixes:

1. Revert back to the old handcrafted path.

2. If the external application directory isn't available, automatically fall back to saving and fetching from the internal storage instead. Maybe move the data across when the external media is mounted.

2. Change the calling code to check for `null` and abort the download with an error message in that case, or prompt the user to select a valid download location.

Option 1 seems to be the simplest, especially keeping in mind the fact that this change is technically out of scope here.

1zaman replied:
Note that you're doing this multiple times in this commit, with the `getParent()` call always having the potential of throwing a `NullPointerException`, so all the instances will need to be addressed together.

miankhalid replied:
Option 1 might be the simplest but that won't solve the issue. Plus, this change might seem out-of-scope but since I have to run migrations and do some renamings, so, I am actually stuck with fixing this.

I do have a solution in mind, which is to standardize the usage of app's external storage access using the FileUtil class, in which I'll add a function which will return the app's external storage directory. Something like this:

````java
    @NonNull
    public static File getAppExternalDir(@NonNull Context context) {
        File externalFilesDir = context.getExternalFilesDir(null);
        if (externalFilesDir != null) {
            return externalFilesDir.getParentFile();
        } else {
            File android = new File(Environment.getExternalStorageDirectory(), "Android");
            File downloadsDir = new File(android, "data");
            File packageDir = new File(downloadsDir, context.getPackageName());
            packageDir.mkdirs();
            return packageDir;
        }
    }
````

miankhalid replied:
I kinda had this thing in mind, while i was using the `context.getExternalFilesDir(null)` function, knowing that it **can return null**, but somehow forgot to implement it.

I guess somethings are best left for the reviewers to pick up 😉

1zaman replied:
Your solution seems to be in line with option 2, except that instead of falling back to the internal storage, you're instead constructing an invalid path to an external storage which would not at the time be actually mounted.

miankhalid replied:
The mount case won't be a problem imo, cuz the methods we are actually using get us access to storage that is marked as external storage by the device from within its internal storage. From what I know the mountable external sd card i.e. micro sd card can't be accessed from these functions and varies from device to device and manufacturers.

ref: http://stackoverflow.com/a/5695129/1402616

1zaman replied:
If you reread your own reference carefully you'll note that it says that the external storage may be either removable media or a specific section of it's non-removable flash storage, as determined by the device manufacturer.

miankhalid replied:
So whats the best way to do what we are trying to do here ?

miankhalid replied:
@bguertin any thoughts on this matter ?

1zaman replied:
I am in favor of option 2.

1zaman replied:
As for what to do if the external storage is not available when updating the database, I suppose there are three options for handling this as well:

1. Don't migrate the videos at all.
2. Listen for the `ACTION_MEDIA_MOUNTED` broadcast, and do the migration at that point. This would probably be the 'proper' course of action (to make everything available as soon as technically possible), but it would also be the most complex logic.
3. Fail the database upgrade until such a time as when the external storage may be mounted. This may leave the app in an unusable state if the external storage is not mounted for an extended period of time, so it's probably not a good idea.

bguertin replied:
As far as writing to external storage, I think we should ONLY use `getExternalFilesDir`.

Like @1zaman mentioned, even if we could construct the path using another method, writing to that path would likely fail. So I would go with option (3) @1zaman mentioned further up: "Change the calling code to check for null and abort the download with an error message"

As for how to handle the upgrade, we also need to account for the case where we only did part of the migration, as the migration is non-transactional.

In either case, we must either:
1) Prevent the user from using the app / certain features until the migration completes (that doesn't sound great)
2) Allow the app / feature to continue to work even if the migration fails (support having the data in either location)

In either case, we can also decide when to attempt migration, either on app-open, user-input, and/or when we see `ACTION_MEDIA_MOUNTED`, etc.

1zaman replied:
> As for how to handle the upgrade, we also need to account for the case where we only did part of the migration, as the migration is non-transactional.

Hmm, so thinking about this, we could actually make this transactional, by performing the whole upgrade in a transaction, and then only marking it as successful after the directory is successfully renamed (which is an atomic operation as well). Otherwise, the whole transaction would be rolled back. If the external files directory path is inaccessible from the start, then the transaction would be aborted as well.

Note that `SQLiteOpenHelper` already performs the upgrade inside a transaction. The Android SQLite framework allows nested transactions, which are actually performed as one single transaction. If any nested transaction is unsuccessful, then the root transaction is rolled back. This should have the implication the database should be accessible in the original version (since the version upgrade would be rolled back as well) and schema until the migration is completed. This would have one caveat though, which is that if we're upgrading from an even older version, then the upgrade wouldn't be performed at all, since it's all done in one transaction. It's been more than a year since the last database update though, so probably there aren't many people at these old versions who would be updating now.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/module/prefs/UserPrefs.java:
```diff
@@ -51,20 +52,21 @@ public boolean isDownloadOverWifiOnly() {
+        File usersVidDir = new File(videosDir, Sha1Util.SHA1(profile.username));
+        if (!usersVidDir.exists()) {
+            usersVidDir.mkdirs();
+        }
```
Again, the `exists()` check is not needed here.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/module/prefs/UserPrefs.java:
```diff
@@ -51,20 +52,21 @@ public boolean isDownloadOverWifiOnly() {
+            File noMediaFile = new File(usersVidDir, ".nomedia");
+            if (!noMediaFile.exists()) {
+                noMediaFile.createNewFile();
+            }
```
The `exists()` check is not needed here either. The `createNewFile()` method also simply returns `false` if the file already exists.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/module/prefs/UserPrefs.java:
```diff
@@ -51,20 +52,21 @@ public boolean isDownloadOverWifiOnly() {
-        File edxDir = new File(packDir, profile.username);
-        edxDir.mkdirs();
+        File videosDir = new File(context.getExternalFilesDir(null).getParent(), AppConstants.Folders.VIDEOS);
+        File usersVidDir = new File(videosDir, Sha1Util.SHA1(profile.username));
```
I don't think 'videos' needs to be shortened to 'vid' in the variable name.

miankhalid replied:
I am definitely not refactoring that, the word/abbreviation **`vid`** is way cooler than `videos`.

miankhalid replied:
But i can rename it to `usersVidsDir`, if plurality is what you had in mind 😉

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/module/prefs/UserPrefs.java:
```diff
@@ -51,20 +52,21 @@ public boolean isDownloadOverWifiOnly() {
         }
-        
-        return edxDir;
+
```
Extra empty line before `return` statement.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/player/TranscriptManager.java:
```diff
@@ -30,24 +30,18 @@
+        transcriptFolder = new File(videosDir, AppConstants.Folders.SUBTITLES);
+        if (!transcriptFolder.exists()) {
+            transcriptFolder.mkdirs();
         }
```
Again, this `exists()` check is not necessary.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/util/AppConstants.java:
```diff
@@ -11,4 +10,9 @@
 
     public static final double MILLISECONDS_PER_SECOND = 1000.00;
+
+    public static final class Folders {
```
Maybe rename this to `Directories`?

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/util/AppConstants.java:
```diff
@@ -11,4 +10,9 @@
+    public static final class Folders {
+        public static final String VIDEOS = "videos";
+        public static final String SUBTITLES = "subtitles";
+    }
```
Could we have some documentation on the `Folders` class, and the constants?

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/util/SecurityUtil.java:
```diff
@@ -0,0 +1,50 @@
+        try {
+            PackageInfo packageInfo = packageManager.getPackageInfo(context.getPackageName(), 0);
+            File appDir = new File(packageInfo.applicationInfo.dataDir);
+            if (appDir.exists() && appDir.isDirectory()) {
```
Being a directory already implies existence, so the extra `exists()` check here is redundant.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/util/SecurityUtil.java:
```diff
@@ -0,0 +1,50 @@
+            PackageInfo packageInfo = packageManager.getPackageInfo(context.getPackageName(), 0);
+            File appDir = new File(packageInfo.applicationInfo.dataDir);
+            if (appDir.exists() && appDir.isDirectory()) {
+                File[] children = appDir.listFiles();
```
Note that you can simply get the list of files, and check if the returned array is a `null` value, instead of having to check the existence and directory status of the `File` as well. This would also make this an atomic operation, and not subject to race conditions like your current conditional logic is.

1zaman replied:
If the directory was deleted between the evaluation of the `if` condition and the request for listing, then it would return `null` anyway. Therefore the check doesn't guarantee anything, and should be replaced by a `null` check on the result of the `listFiles()` method.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/util/SecurityUtil.java:
```diff
@@ -32,5 +33,16 @@ public static void clearUserData(Context context) {
+        File externalDir = new File(context.getExternalFilesDir(null).getParent());
+        if (externalDir.exists() && externalDir.isDirectory()) {
+            File[] children = externalDir.listFiles();
+            for (int size = children.length, i = 0; i < size; i++) {
```
All the comments from the previous operation apply here as well.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/module/db/DbStructure.java:
```diff
@@ -8,9 +8,12 @@
+    /**
+     * Updated to Version 4 to add flag to indicate that video is only available for web
+     * Updated to Version 5 to create a new table to record learning history for assessment
+     * Updated to Version 6 to swap every occurrence of username field to its SHA1 counterpart
```
'counterpart' can be replaces with 'hash' for more clarity, I suppose.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/util/SecurityUtil.java:
```diff
@@ -0,0 +1,50 @@
+import java.io.File;
+import java.util.Collections;
+
+public abstract class SecurityUtil {
```
This class shouldn't be `abstract`, just non-instantiable.

1zaman replied:
Also, could we have some javadocs on the scope of this class, and the `clearUserData()` method?

1zaman commented on VideoLocker/src/test/java/org/edx/mobile/test/UtilTests.java:
```diff
@@ -32,4 +34,10 @@ public void testHostAndUrls() throws Exception {
     }
+
+    @Test
+    public void testSha1Encryption() {
```
I think it would be better to have separate test suites for each utility class for better organization (both for the reader and the test code itself). In which case the `convertToHex()` method should be tested as well I guess.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/module/prefs/PrefManager.java:
```diff
@@ -355,6 +357,10 @@ public void setLastCourseStructureFetch(String courseId, long timestamp) {
         public static final String USER_PREF = "pref_user";
+
+        public static String[] getAll() {
+            return new String[]{LOGIN, WIFI, VIDEOS, FEATURES, APP_INFO, USER_PREF};
```
Maybe add some whitespace separation in the array definition?

```` java
new String[] { LOGIN, ... };
````

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/module/db/impl/IDatabaseImpl.java:
```diff
@@ -759,6 +757,30 @@ public Boolean isAnyVideoDownloadingInSubSection(final DataCallback<Boolean> cal
+                @Override
+                public void onResult(List<String> result) {
+                    callback.sendResult(result);
+                }
```
Why do you need a wrapper for the data callback? The executer already calls `sendResult()` instead of invoking the `onResult()` callback method directly.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/task/RestoreVideosCacheDataTask.java:
```diff
@@ -0,0 +1,37 @@
+
+    @Override
+    protected void onSuccess(Void aVoid) throws Exception {
+        super.onSuccess(aVoid);
```
There is no `super` implementation of success, so calling through to it is redundant.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/module/prefs/PrefManager.java:
```diff
@@ -413,6 +413,9 @@ public void setIsVideosCacheRestored(boolean restored) {
     }
 
+    /**
+     * Clears all the shared preferences being used in app.
```
Grammar fix: Clears all the shared preferences **that are** being used in app.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/util/FileUtil.java:
```diff
@@ -48,8 +50,9 @@
-        if (exceptions.contains(fileOrDirectory.getName())) return;
+    public static void deleteRecursive(@NonNull File fileOrDirectory,
+                                       @Nullable List<String> exceptions) {
+        if (exceptions != null && exceptions.contains(fileOrDirectory.getName())) return;
```
I think it would be better to not have the `exceptions` parameter declared as `Nullable`, and force the caller to send an empty `List` instead, making for more concise code. Note that there is a shared constant immutable empty `List` available at `Collections.EMPTY_LIST` for this kind of scenario, so this wouldn't need to have the cost of temporarily created object instantiation either. If there are lost of callers that don't have exceptions, then there could be a convenience duplicate method without this parameter, which might call through to this method with the empty list itself.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/module/prefs/PrefManager.java:
```diff
@@ -392,6 +400,7 @@ public void setLastCourseStructureFetch(String courseId, long timestamp) {
         public static final String AppSettingNeedSyncWithParse = "AppSettingNeedSyncWithParse";
         public static final String UserPrefVideoModel = "UserPrefVideoModel";
         public static final String LAST_COURSE_STRUCTURE_FETCH = "LastCourseStructureFetch";
+        public static final String VIDEOS_CACHE_RESTORED = "VideosCacheRestored";
```
Should this be documented here. I see that the keys (and the accessor and setter methods) aren't documented here, but this may be a temporary key so it may be worth explicitly documenting it's purpose and scope here to aid in it's potential future removal.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/view/MyCourseListTabFragment.java:
```diff
@@ -32,13 +35,19 @@
+        PrefManager.UserPrefManager prefs = new PrefManager.UserPrefManager(MainApplication.application);
+        if (!prefs.isVideosCacheRestored()) {
+            new RestoreVideosCacheDataTask(MainApplication.application).execute();
+        }
```
This logic could be moved to a static method in `RestoreVideosCacheDataTask`, e.g. `RestoreVideosCacheDataTask.executeInstanceIfNeeded()` for better symmetry (i.e. it sets the shared preference value so might as well have it check the value as well). But it's up to you; this is fine as well.

1zaman commented:
OK, my review is completed.

miankhalid commented:
@1zaman I've addressed your review comments.

Ready for another review @1zaman, @mdinino, @bguertin.

These are the commits that need review: 07ae69d & c6b3478.

1zaman commented on OpenEdXMobile/src/test/java/org/edx/mobile/test/Sha1UtilTest.java:
```diff
@@ -0,0 +1,20 @@
+
+public class Sha1UtilTest extends BaseTestCase {
+    @Test
+    public void testSha1Encryption() {
```
Would be nice (but not essential) to have some javadocs for this test suite that link back to the tested class/methods.

miankhalid replied:
Not needed. The names are self-explanatory.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/util/SecurityUtil.java:
```diff
@@ -13,20 +12,31 @@
 
-public abstract class SecurityUtil {
+/**
+ * Utility class that deals with the security of a user's personal information data.
```
I suggest replacing "that deals" with "for dealing" or just "dealing".

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/util/SecurityUtil.java:
```diff
@@ -13,20 +12,31 @@
+public class SecurityUtil {
     private static final Logger logger = new Logger(SecurityUtil.class);
 
+    private SecurityUtil() {
```
Might as well throw an exception here as well, to make this impossible to instantiate successfully.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/util/SecurityUtil.java:
```diff
@@ -13,20 +12,31 @@
+    }
+
+    /**
+     * Clears the app's data directory, external storage directory and shared preferences with the
```
There should be a comma before "and" and "with the".

miankhalid replied:
> comma before "and"

there's never a comma before "and" in these situations e.g.

- I'll do this, this and that. (correct)
- I'll do this, this, and that. (incorrect)

miankhalid replied:
Never read that **thing**, was never taught that. Its one of those corruption by the people on the other side of the pond that defies logic, physics, chemistry, sense and what not.

Yeah! you read it right, I did'nt use the serial comma again before that **and**.

Let it be too strong. 🤓

1zaman replied:
s/did'nt/didn't #spellingnazi

miankhalid replied:
<img src="http://vignette3.wikia.nocookie.net/brutallegend/images/8/8c/You_Don't_Say.jpg/revision/latest?cb=20120601161805" />

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/util/SecurityUtil.java:
```diff
@@ -13,20 +12,31 @@
+
+    /**
+     * Clears the app's data directory, external storage directory and shared preferences with the
+     * exceptions of downloaded videos and app's database.
```
The "app's" in "app's database" is redundant; maybe change to "video database" to put it in the proper context.

1zaman commented on OpenEdXMobile/src/main/java/org/edx/mobile/module/db/impl/DbHelper.java:
```diff
@@ -77,98 +77,102 @@ private void createAssessmentTable(SQLiteDatabase db){
 
     @Override
     public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
+        db.beginTransaction();
```
This scoping is not correct. You're starting the transaction from the start, but only marking it as successful (unconditionally) in the specific upgrade to version 6. This would mean that the transaction would fail on subsequent future upgrades.

What you need to do is to only begin (and end) the nested transaction in the version 6 upgrade specifically, and only mark it as successful if the migration is truly successful (i.e. the external storage is mounted, and the directory move and database upgrades are successful.

1zaman commented on OpenEdXMobile/src/main/java/org/edx/mobile/player/TranscriptManager.java:
```diff
@@ -23,17 +25,12 @@
-    private File transcriptFolder;
-    private Context context;
     private final Logger logger = new Logger(getClass().getName());
+    private Context context;
```
This should be declared as `final`.

1zaman commented on OpenEdXMobile/src/main/java/org/edx/mobile/player/TranscriptManager.java:
```diff
@@ -55,8 +55,11 @@ public boolean has(String url) {
      */
     public void put(String url, String response) throws IOException {
+        final File transcriptDir = getTranscriptDir();
+        if (transcriptDir == null) throw new IOException("Transcript directory doesn't exist");
```
I suggest changing "doesn't exist" to "not found" or "nor available" in all these methods.

1zaman commented on OpenEdXMobile/src/main/java/org/edx/mobile/util/FileUtil.java:
```diff
@@ -83,17 +83,9 @@ public static void deleteRecursive(@NonNull File fileOrDirectory,
-            packageDir.mkdirs();
-            return packageDir;
-        }
+        return (externalFilesDir != null ? externalFilesDir.getParentFile() : null);
```
The parentheses around the returned expression aren't necessary.

miankhalid replied:
2 Reasons they are there:

- To improve readability.
- I favor this way of encapsulating a ternary operation.

1zaman commented on OpenEdXMobile/src/main/java/org/edx/mobile/module/prefs/PrefManager.java:
```diff
@@ -280,7 +285,7 @@ public void setIsVideosCacheRestored(boolean restored) {
 
     /**
-     * Clears all the shared preferences being used in app.
+     * Clears all the shared preferences that are being used in app.
```
Another small grammer fix: it should be "in **the** app". Also, "being" could be removed without losing any essential meaning here, I suppose.

miankhalid replied:
[You are contradicting yourself.](https://github.com/edx/edx-app-android/pull/797#r76391499)

1zaman replied:
Not really; these can be considered incremental improvements 😜

miankhalid replied:
(╯°□°)╯︵ ┻━┻

miankhalid commented on OpenEdXMobile/src/main/java/org/edx/mobile/module/db/impl/IDatabaseImpl.java:
```diff
@@ -757,26 +757,14 @@ public Boolean isAnyVideoDownloadingInSubSection(final DataCallback<Boolean> cal
 
     @Override
-    public List<String> getUniqueCourseIdsForDownloadedVideos(final DataCallback<List<String>> callback) {
+    public List<String> getUniqueCourseIdsForDownloadedVideos(@Nullable final DataCallback<List<String>> callback) {
```
Since, I've just touched this method in my PR, so, adding the `@Nullable` in the interface declaration as well.

miankhalid commented on OpenEdXMobile/src/main/java/org/edx/mobile/module/prefs/PrefManager.java:
```diff
@@ -266,6 +266,11 @@ public void setIsVideosCacheRestored(boolean restored) {
         public static final String UserPrefVideoModel = "UserPrefVideoModel";
         public static final String LAST_COURSE_STRUCTURE_FETCH = "LastCourseStructureFetch";
+        /**
+         * For downloaded videos to appear in order on the My Videos screen, we need
```
"in order" is the whole reason why we need the videos' courses data cached. Cuz the order in which the videos have been downloaded may differ from the order they appear in courseware.

miankhalid commented:
Ok then, @1zaman I've fixed all the review points u mentioned or I had left previously. Hopefully we'll get them approved this time 😉 

The commits to review are as follows:

- Missed review points 5065d40
- Final review changes 12cf6fe
- Separate function for manual logout on button tap f3afb2b
- Singleton issues fix for CacheManager 6ad5186 (This was one pesky issue I had real hard time finding, causing everything to break upon relogin)

1zaman commented on OpenEdXMobile/src/main/java/org/edx/mobile/module/db/impl/DbHelper.java:
```diff
@@ -77,73 +80,86 @@ private void createAssessmentTable(SQLiteDatabase db){
+                                                    hashedUsername)).toString());
+                                    if (!((newDir.mkdirs() || newDir.exists()) &&
+                                            previousDir.renameTo(newDir))) {
+                                        continue;
```
Hmm, what should we do with the database entry if this situation is somehow encountered. Maybe just delete the entry here?

miankhalid replied:
hmm, yeah, i guess, we should delete it.

1zaman commented on OpenEdXMobile/src/main/java/org/edx/mobile/module/db/impl/DbHelper.java:
```diff
@@ -69,52 +80,118 @@ private void createAssessmentTable(SQLiteDatabase db){
-        }catch(Exception e){
-            logger.error(e);
+                logger.debug("Database upgraded from " + oldVersion + " to " + newVersion);
+                db.setTransactionSuccessful();
```
You're [still][1] marking it as successful unconditionally. This should be moved to inside the `if (externalAppDir != null)` condition at the end (i.e. just 2-3 lines above before ending it.

[1]: https://github.com/edx/edx-app-android/pull/797#discussion_r81094549

miankhalid replied:
oops 😞

miankhalid commented on OpenEdXMobile/src/main/java/org/edx/mobile/module/db/impl/DbHelper.java:
```diff
@@ -77,73 +80,86 @@ private void createAssessmentTable(SQLiteDatabase db){
+                                        TextUtils.join("/", Arrays.<CharSequence>asList(
+                                                externalAppDir.getAbsolutePath(),
+                                                AppConstants.Directories.VIDEOS,
+                                                hashedUsername)).toString());
```
I don't see how adding a `/` @ end will make it any different or robust ?

miankhalid replied:
Lets just delete the row once we reach the next `if`.

1zaman replied:
Hmm, OK. Which reminds me that you should also change the `previousDir` path to use `previousAppDirPath` instead of `externalAppDir`.

miankhalid replied:
Aren't we already doing that, I remember doing it.

miankhalid replied:
oops, I'm not.

1zaman replied:
Actually, this seems to be a different check, since if the desired pattern is not found in the original filepath, the directory might still exist, and will thus be skipped over by that checked. Better to just add this check here as well.

miankhalid replied:
ok.

miankhalid commented:
@1zaman changes done, hopefully the last time 🙏 🙏 😝

1zaman commented:
👍 All my comments seem to be addressed, but I just made [another][1] one 😜 If you would address that as well, then I'll finally approve this.

[1]: https://github.com/edx/edx-app-android/pull/797#discussion_r82497767

miankhalid commented:
@1zaman ok, np. If thats the last thing, should i just do that and squash the commits ?

1zaman commented:
Sure, you can squash the commits now.
